iroh/src/node/builder.rs

```diff
@@ -715,6 +690,229 @@ where
     }
 }
 
+/// A node that is initialized but not yet spawned.
+///
+/// This is returned from [`Builder::build`] and may be used to register custom protocols with
+/// [`Self::accept`]. It provides access to the services which are already started, the node's
+/// endpoint and a client to the node.
+///
+/// Note that the client returned from [`Self::client`] can only be used after spawning the node,
+/// until then all RPC calls will time out.
+#[derive(derive_more::Debug)]
+pub struct UnspawnedNode<D, E> {
+    inner: Arc<NodeInner<D>>,
+    client: crate::client::MemIroh,
+    internal_rpc: FlumeServerEndpoint<RpcService>,
+    rpc_endpoint: E,
+    protocols: ProtocolMap,
+    #[debug("callback")]
+    gc_done_callback: Option<Box<dyn Fn() + Send>>,
+    gc_policy: GcPolicy,
+}
+
+impl<D: iroh_blobs::store::Store, E: ServiceEndpoint<RpcService>> UnspawnedNode<D, E> {
+    /// Register a protocol handler for incoming connections.
+    ///
+    /// Use this to register custom protocols onto the iroh node. Whenever a new connection for
+    /// `alpn` comes in, it is passed to this protocol handler.
+    ///
+    /// See the [`Protocol`] trait for details.
+    ///
+    /// Example usage:
+    ///
+    /// ```rust
+    /// # use std::sync::Arc;
+    /// # use anyhow::Result;
+    /// # use futures_lite::future::Boxed as BoxedFuture;
+    /// # use iroh::{node::{Node, Protocol}, net::endpoint::Connecting, client::MemIroh};
+    /// #
+    /// # #[tokio::main]
+    /// # async fn main() -> Result<()> {
+    ///
+    /// const MY_ALPN: &[u8] = b"my-protocol/1";
+    ///
+    /// #[derive(Debug)]
+    /// struct MyProtocol {
+    ///     client: MemIroh
+    /// }
+    ///
+    /// impl Protocol for MyProtocol {
+    ///     fn accept(self: Arc<Self>, conn: Connecting) -> BoxedFuture<Result<()>> {
+    ///         todo!();
+    ///     }
+    /// }
+    ///
+    /// let unspawned_node = Node::memory()
+    ///     .build()
+    ///     .await?;
+    ///
+    /// let client = unspawned_node.client().clone();
+    /// let handler = MyProtocol { client };
+    ///
+    /// let node = unspawned_node
+    ///     .accept(MY_ALPN, Arc::new(handler))
+    ///     .spawn()
+    ///     .await?;
+    /// # node.shutdown().await?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    ///
+    ///
+    pub fn accept(mut self, alpn: &'static [u8], handler: Arc<dyn Protocol>) -> Self {
+        self.protocols.insert(alpn, handler);
+        self
+    }
+
+    /// Return a client to control this node over an in-memory channel.
+    ///
+    /// Note that the client can only be used after spawning the node,
+    /// until then all RPC calls will time out.
```

Comment:
If you grab the client, then do a rpc call, then immediately finish the build, a RPC call might work, right?

So it's not so much that they will time out but that they will hang until the rpc accept loop goes live.

Reply:
Changed the docs.